cargo-shuttle/src/lib.rs

```diff
@@ -1103,17 +1090,101 @@ impl Shuttle {
+                    match shuttle_resource.r#type {
+                        resource::Type::Database(db_type) => {
+                            let config: DbInput = serde_json::from_value(shuttle_resource.config)
+                                .context("deserializing resource config")?;
+                            let res = match config.local_uri {
+                                Some(local_uri) => DatabaseResource::ConnectionString(local_uri),
+                                None => DatabaseResource::Info(
+                                    prov.provision_database(Request::new(DatabaseRequest {
+                                        project_name: service_name.to_string(),
+                                        db_type: Some(db_type.into()),
+                                    }))
+                                    .await?
+                                    .into_inner()
+                                    .into(),
+                                ),
+                            };
+                            shuttle_resources.push(resource::Response {
+                                r#type: shuttle_resource.r#type,
+                                config: serde_json::Value::Null,
+                                data: serde_json::to_value(&res).unwrap(),
+                            });
+                            *r = serde_json::to_vec(&ShuttleResourceOutput {
+                                output: res,
+                                custom: shuttle_resource.custom,
+                            })
+                            .unwrap();
+                        }
+                        resource::Type::Secrets => {
+                            shuttle_resources.push(resource::Response {
+                                r#type: shuttle_resource.r#type,
+                                config: serde_json::Value::Null,
+                                data: serde_json::to_value(secrets.clone()).unwrap(),
+                            });
+                            *r = serde_json::to_vec(&ShuttleResourceOutput {
+                                output: secrets.clone(),
+                                custom: shuttle_resource.custom,
+                            })
+                            .unwrap();
+                        }
+                        resource::Type::Persist => {
+                            shuttle_resources.push(resource::Response {
+                                r#type: shuttle_resource.r#type,
+                                config: serde_json::Value::Null,
+                                data: serde_json::Value::Null,
+                            });
+                        }
+                        resource::Type::Container => {
+                            let config = serde_json::from_value(shuttle_resource.config)
+                                .context("deserializing resource config")?;
+                            let res = prov.start_container(config).await?;
```

Comment:
This loop currently does a number of things:
1. It adds resources to `shuttle_resources`
2. It sometimes changes the resource
3. It sometimes provisions a resource

It is currently hard to reason about all of these. Can you somehow use something like a functional style to better separate these three tasks